File: cs/1_4_builder.cs
using System.Collections.Generic;
using System;

namespace Creational
{
    /* 
        Приклад шаблону Будівельник.
        Визначено такі етапи побудови:
        1. Додавання частини
        2. Додавання часової мітки
        3. Зміна імені
    */ 
    namespace Builder
    {
        class Product
        {
            private List<string> parts = new List<string>();
            public string Name = "No name";
            public void Add(string part)
            {
                this.parts.Add(part);
            }
            
            public override string ToString ()
            {
                string str = string.Empty;
                foreach (string part in this.parts)
                {
                    str += $"\t{part},\n ";
                }
                return $"Product <{this.Name}> parts: \n {str}";
            }
        }
        
        interface IBuilder
        {
            IBuilder AddPart(string part);
            IBuilder SetDateStamp();
            IBuilder SetName(string name);
            Product GetProduct();
            void Reset();
        }

        class Builder : IBuilder
        {
            protected Product product  = new Product();

            public void Reset()
            {
                this.product = new Product();
            }
            public IBuilder SetName(string name)
            {
                this.product.Name = name;
                return this; //для можливості побудови ланцюга виклику методів 
            }

            public IBuilder AddPart (string part)
            {
                this.product.Add(part);
                return this;
            }

            public virtual IBuilder SetDateStamp ()
            {
                this.product.Add($"Date stamp: {DateTime.Now.ToString()}");
                return this;
            }

            public Product GetProduct()
            {
                Product result = this.product;
                this.Reset();
                return result;
            }
        }

        class OtherBuilder : Builder {
            public override IBuilder SetDateStamp ()
            {
                this.product.Add($"{DateTime.Now.ToString()}");
                return this;
            }
        }
        /* клас Директор реалізує шаблон Фабричний метод 
        та генерує деякі пресети з використанням Будівельника
        */ 
        class Director
        {
            public IBuilder builder;
            public Director(IBuilder builder)
            {
                this.builder = builder;
            }

            public Product Empty()
            {
                this.builder.Reset();
                return this.builder.GetProduct();
            }

            public Product BuildFromParts (string[] parts)
            {
                this.builder.Reset();
                foreach (string part in parts)
                {
                    this.builder.AddPart(part);
                }
                return this.builder.GetProduct();
            }

            public Product Example()
            {
                this.builder.Reset();
                return this.builder
                        .SetName("Example")
                        .AddPart("Part One")
                        .AddPart("Part Two")
                        .SetDateStamp()
                        .AddPart("Part Three")
                        .GetProduct();
            }
        }
    }
}

File: python/patterns/_1_4_builder.py
from __future__ import annotations
from typing import Protocol, Self
from datetime import datetime, timezone


class BuildProduct:
    def __init__(self):
        self._parts: list[str] = []
        self.name: str = "No name"

    def __repr__(self) -> str:
        return f"Product {self.name}, parts:\n {','.join(self._parts)}"

    def add_part(self, part: str) -> None:
        self._parts.append(part)


class Builder(Protocol):
    def add_part(self, part: str) -> Self: ...
    def set_stamp(self) -> Self: ...
    def set_name(self, name: str) -> Self: ...
    def get_product(self) -> BuildProduct: ...
    def reset(self) -> None: ...


class SomeBuilder:#(Builder):
    def __init__(self) -> None:
        super().__init__()
        self._product: BuildProduct = BuildProduct()

    def add_part(self, part: str) -> Self:
        self._product.add_part(part)
        return self

    def set_stamp(self) -> Self:
        self._product.add_part(f"Date stamp: {datetime.now()}")
        return self

    def set_name(self, name: str) -> Self:
        self._product.name = name
        return self

    def get_product(self) -> BuildProduct:
        result: BuildProduct = self._product
        self.reset()
        return result

    def reset(self) -> None:
        self._product = BuildProduct()


class OtherBuilder(SomeBuilder):
    def set_stamp(self) -> Self:
        self._product.add_part(f"<{datetime.now(timezone.utc)}>")
        return self


class Director:
    def __init__(self, builder: Builder) -> None:
        self.builder: Builder = builder
        self.builder.reset()

    def build_empty_product(self) -> BuildProduct:
        self.builder.reset()
        return self.builder.get_product()

    def build_from_parts(self, parts: list[str]) -> BuildProduct:
        for part in parts:
            self.builder.add_part(part)
        return self.builder.get_product()

    def build_example(self) -> BuildProduct:
        return self.builder\
            .set_name("example")\
            .add_part("part one")\
            .add_part("part two")\
            .set_stamp()\
            .add_part("part three")\
            .get_product()
